Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximated_centroidal_Voronoi_diagram_remeshing.h

```diff
@@ -0,0 +1,1385 @@
+                                             get_property_map(CGAL::vertex_point, tmesh));
+
+  // get curvature related parameters
+#ifdef CGAL_ACVD_DOES_NOT_USE_INTERPOLATED_CORRECTED_CURVATURES
+  static_assert(is_default_parameter<NamedParameters, internal_np::gradation_factor_t>::value, "gradation_factor option is disabled");
+#endif
+  const FT gradation_factor = choose_parameter(get_parameter(np, internal_np::gradation_factor), 0);
+
+  // using QEM ?
+  bool use_postprocessing_qem = choose_parameter(get_parameter(np, internal_np::use_postprocessing_qem), false);
+  const bool use_qem_based_energy = choose_parameter(get_parameter(np, internal_np::use_qem_based_energy), false);
+  if (use_qem_based_energy)
+    use_postprocessing_qem = false;
+
+#ifndef CGAL_ACVD_DOES_NOT_USE_INTERPOLATED_CORRECTED_CURVATURES
+  Vertex_principal_curvatures_and_directions_map vpcd_map;
+
+  // if adaptive clustering
+  if (gradation_factor > 0)
+  {
+    if constexpr (is_default_parameter<NamedParameters, internal_np::vertex_principal_curvatures_and_directions_map_t>::value)
+    {
+      vpcd_map = get(CGAL::dynamic_vertex_property_t<Principal_curvatures_and_directions<GT>>(), tmesh);
+      interpolated_corrected_curvatures(tmesh, parameters::vertex_principal_curvatures_and_directions_map(vpcd_map)
+                                                          .vertex_point_map(vpm).geom_traits(gt));
+    }
+    else
+      vpcd_map = get_parameter(np, internal_np::vertex_principal_curvatures_and_directions_map);
+  }
+#endif
+
+  CGAL_precondition(CGAL::is_triangle_mesh(tmesh));
+
+  const double vertex_count_ratio = choose_parameter(get_parameter(np, internal_np::vertex_count_ratio), 0.1);
+  // TODO: (possible optimization to save memory) if qem is not used use std::conditionnal to store and compute less things in QEMClusterData
+  // TODO: (possible optimization to save memort) use a std::array to store quadrics and assemble the Eigen matric only at the end
+
+  std::size_t nb_vertices = vertices(tmesh).size();
+
+  // For remeshing, we might need to subdivide the mesh before clustering
+  // This should always hold: nb_clusters <= nb_vertices * vertex_count_ratio
+  // So do the following until the condition is met
+  if (gradation_factor == 0 && nb_clusters > nb_vertices * vertex_count_ratio)
+  {
+    std::vector<FT> lengths;
+    lengths.reserve(num_edges(tmesh));
+    FT cum = 0;
+    int nbe = 0;
+    for (edge_descriptor e : edges(tmesh))
+    {
+      lengths.push_back(edge_length(halfedge(e, tmesh), tmesh, parameters::vertex_point_map(vpm).geom_traits(gt)));
+      cum += lengths.back();
+      ++nbe;
+    }
+    FT threshold = 3. * cum / nbe;
+
+    std::vector<std::pair<edge_descriptor, double>> to_split;
+    int ei = -1;
+    for (edge_descriptor e : edges(tmesh))
+    {
+      FT l = lengths[++ei];
+      if (l >= threshold)
+      {
+        double nb_subsegments = std::ceil(l / threshold);
```

Comment:
yet another `std::ceil()`

Reply:
a class model of `Kernel`, with `GT::FT` being either `float` or `double`.